desc/objectives/_coils.py

```diff
@@ -1267,6 +1283,197 @@ def compute(self, field_params, constants=None):
+                rho=np.array([1.0]),
+                M=2 * surface.M,
+                N=2 * surface.N,
+                NFP=surface.NFP,
+                sym=False,
+            )
+            self._eval_grid = eval_grid
+        else:
+            eval_grid = self._eval_grid
+
+        self._data_keys = ["R", "Z", "n_rho", "phi", "|e_theta x e_zeta|"]
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        self._dim_f = eval_grid.num_nodes
+
+        w = eval_grid.weights
+        w *= jnp.sqrt(eval_grid.num_nodes)
+
+        eval_profiles = get_profiles(self._data_keys, obj=surface, grid=eval_grid)
+        eval_transforms = get_transforms(self._data_keys, obj=surface, grid=eval_grid)
+        eval_data = compute_fun(
+            surface,
+            self._data_keys,
+            params=surface.params_dict,
+            transforms=eval_transforms,
+            profiles=eval_profiles,
+        )
+
+        self._constants = {
+            "field": self._field,
+            "field_grid": self._field_grid,
+            "quad_weights": w,
+            "eval_data": eval_data,
+            "eval_transforms": eval_transforms,
+            "eval_profiles": eval_profiles,
+        }
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        if self._normalize:
+            scales = compute_scaling_factors(surface)
+            Bscale = 1.0  # surface has no inherent B scale
+            self._normalization = Bscale * scales["R0"] * scales["a"]
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params_1, params_2=None, constants=None):
+        """Compute boundary force error.
+
+        Parameters
+        ----------
+        params_1 : dict
+            Dictionary of the surface's degrees of freedom.
+        params_2 : dict
+            Dictionary of the external field's degrees of freedom, only provided if
+            if field_fixed=False.
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        f : ndarray
+            Bnorm on the QFM surface from the external field
+
+        """
+        if constants is None:
+            constants = self.constants
+        field_params = params_2 if not self._field_fixed else None
+        surf_params = params_1
+
+        eval_data = compute_fun(
+            self._surface,
+            self._data_keys,
+            surf_params,
+            constants["eval_transforms"],
+            constants["eval_profiles"],
+        )
+        x = jnp.array([eval_data["R"], eval_data["phi"], eval_data["Z"]]).T
+        if field_params is None:
+            field_params = constants["field"].params_dict
+        B_ext = constants["field"].compute_magnetic_field(
```

Comment:
A Python question. Is `constants["field"]` a copy of the field object at the time of build (so actually a constant) or a pointer to the field object such that when the object changes the `constants["field"]` also changes? If it is the former, then the compute function kind of assumes the field is always fixed?

Reply:
Probably, this implementation is correct but still a bit confused by something inside `constants` to be changing...

Reply:
Okok, I check it is a pointer not a copy

Reply:
It is a pointer BUT also it would not matter if it was a copy, here we only are using its compute_magnetic_field method to use the internal algorithms, we pass in explicitly `field_params` with the current DOFs of the field, which then get used in computing the magnetic field.

Reply:
Hmm ok, thanks